test/local-hub/external_storage_backend_test.go

```diff
@@ -0,0 +1,580 @@
+			},
+			dataInExternalBackend: map[string]interface{}{
+				"updated-value": "context 0",
+			},
+			locked:                  "",
+			lockedInExternalBackend: "",
+		},
+		{
+			alias: ptr.String("child"),
+			backendContext: map[string]interface{}{
+				"updated": "context 1",
+			},
+			dataInGQL: map[string]interface{}{
+				"updated-value": "context 1",
+			},
+			dataInExternalBackend:   nil,
+			locked:                  "",
+			lockedInExternalBackend: "",
+		},
+		{
+			alias:          ptr.String("original"),
+			backendContext: nil,
+			dataInGQL: map[string]interface{}{
+				"name": "Anakin Skywalke",
+			},
+			dataInExternalBackend: map[string]interface{}{
+				"name": "Anakin Skywalke",
+			},
+			locked:                  "",
+			lockedInExternalBackend: "",
+		},
+		{
+			alias: ptr.String("parent"),
+			backendContext: map[string]interface{}{
+				"provider": "dotenv",
+			},
+			dataInGQL: map[string]interface{}{
+				"name": "Darth Vader",
+			},
+			dataInExternalBackend: map[string]interface{}{
+				"name": "Darth Vader",
+			},
+			locked:                  "",
+			lockedInExternalBackend: "",
+		},
+	})
+}
+
+func registerExternalDotenvStorage(ctx context.Context, t *testing.T, cli *local.Client, srvAddr string) (gqllocalapi.CreateTypeInstanceOutput, func()) {
+	t.Helper()
+
+	ti, err := cli.CreateTypeInstances(ctx, fixExternalDotenvStorage(srvAddr))
+	require.NoError(t, err)
+	require.Len(t, ti, 1)
+	dotenvHubStorage := ti[0]
+
+	return dotenvHubStorage, func() {
+		_ = cli.DeleteTypeInstance(ctx, dotenvHubStorage.ID)
+	}
+}
+
+func fixExternalDotenvStorage(addr string) *gqllocalapi.CreateTypeInstancesInput {
+	return &gqllocalapi.CreateTypeInstancesInput{
+		TypeInstances: []*gqllocalapi.CreateTypeInstanceInput{
+			{
+				CreatedBy: ptr.String("manually"),
+				TypeRef: &gqllocalapi.TypeInstanceTypeReferenceInput{
+					Path:     "cap.type.example.filesystem.storage",
+					Revision: "0.1.0",
+				},
+				Value: storageSpec{
+					URL:         ptr.String(addr),
+					AcceptValue: ptr.Bool(true),
+					ContextSchema: ptr.String(heredoc.Doc(`
+				      {
+				      	"$id": "#/properties/contextSchema",
+				      	"type": "object",
+				      	"properties": {
+				      		"provider": {
+				      			"$id": "#/properties/contextSchema/properties/name",
+				      			"type": "string",
+								  "enum": [
+									"aws_secretsmanager",
+									"dotenv"
+								  ]
+				      		}
+				      	},
+				      	"additionalProperties": true
+				      }`)),
+				},
+			},
+		},
+		UsesRelations: []*gqllocalapi.TypeInstanceUsesRelationInput{},
+	}
+}
+
+type externalData struct {
+	Value    string
+	LockedBy *string
+}
+
+func getDataDirectlyFromStorage(t *testing.T, addr string, details []gqllocalapi.TypeInstance) map[string]externalData {
+	t.Helper()
+	url := strings.Split(addr, ":")
+	assert.Equal(t, len(url), 2)
+	conn, err := grpc.Dial(fmt.Sprintf("[%s]:%s", url[0], url[1]), grpc.WithInsecure())
```

Comment:
Questions:

1. Why do we have square brackets in the URL? What does it do?
2. Can we use the `addr` there directly?